core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,7 +64,23 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // Evaluate partitions in parallel. Partitions of each rdd will be cached by the `partitions`
+  // val in `RDD`.
+  private[spark] lazy val parallelPartitionEval: Boolean = {
```

Comment:
Agreed. The root cause would still be a badly flawed custom implementation, and the right fix to fix that, and I don't see any evidence that this exists in the wild. Do you have a specific example in mind?

Otherwise I personally would not merge this change for this reason. I would support merging the simpler change to parallelize the loop.